### PR TITLE
[MM-54933] Add the ability to @ mention custom groups in group constrained teams and channels

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
There was a bug where you could not @ mention custom groups in group constrained teams, this fixes that issue. I think the ldap and custom group codes needs a complete review, there is a lot going on.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-54933

#### Release Note

```release-note
Fixed a bug where you couldn't @ mention custom groups in group constrained teams and channels
```

### CODE REVIEW

File: server/channels/app/notification.go [hunk 1]
```diff
@@ -1254,6 +1254,18 @@ func (a *App) getGroupsAllowedForReferenceInChannel(channel *model.Channel, team
...
+		var customgroups []*model.Group
+		customgroups, err = a.Srv().Store().Group().GetGroups(0, 0, opts, nil)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to get custom groups")
+		}
+		for _, group := range customgroups {
+			if group.Name != nil {
+				groupsMap[*group.Name] = group
+			}
+		}
```
Comment (hanzei) [inline server/channels/app/notification.go:1268]: Can a customer group have the same name as a channel or team group, i.e. are group names globally unique?

**Review by hanzei — COMMENTED** [review]
Nice work on the tests :+1:

File: server/channels/app/notification.go (hunk 1 above)
Comment (BenCookie95) [reply server/channels/app/notification.go:1268]: Sorry missed this comment, the group names are globally unique

**Review by hanzei — APPROVED** [review]
LGTM :+1:

**Review by JulienTant — APPROVED** [review]
LGTM, thanks @BenCookie95

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Custom groups are now shown even when LDPA sync is enforced for team membership as well as on channels when channel membership is enforced
- When there are no group members present on the team for the custom group a system message gives UI feedback 
- When a member of the custom group is also in the LDPA group and Team sync is enforced, the mention works normally
- Regression tested custom group CRUD operations and ensured mention options are removed when a group is archived
- LDAP sync enforcement on/off works as expected to remove members when sync is run
- Tested LDAP group sync for teams as well as channels  
- Enable Group Mentions for a specific LDAP groups (feature labeled as Beta in the UI) and it works as expected

LGTM!
Note: This issue was a bit specific to find. I'll open a PR shortly to add a Zephyr test case for this. 

Thanks @BenCookie95 👍

**Pull request merged**